# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/WindowContext.h
# Commit: 108d5fb4418e
# Full Hash: 108d5fb4418e78e4dca0a746863e5784c6d80371
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-04-23 09:52:48
# Regressor Bug: 1580565
# File Overlap Count: 3
# Description:
#   Bug 1580565 - Part 4: Use WindowContext to manage BrowsingContext cached status, r=farre
#   
#   The existing infrastructure which stored cached BrowsingContexts on the
#   BrowsingContextGroup was added before WindowContexts were added, and can cause
#   racing issues with partially discarded trees during process switches.
# ==============================================================================

diff -r d8dea951a032 -r 108d5fb4418e docshell/base/WindowContext.h
--- a/docshell/base/WindowContext.h	Wed Apr 22 15:48:20 2020 +0000
+++ b/docshell/base/WindowContext.h	Wed Apr 22 15:48:17 2020 +0000
@@ -7,6 +7,7 @@
 #ifndef mozilla_dom_WindowContext_h
 #define mozilla_dom_WindowContext_h
 
+#include "mozilla/Span.h"
 #include "mozilla/dom/MaybeDiscarded.h"
 #include "mozilla/dom/SyncedContext.h"
 #include "mozilla/net/NeckoChannelParams.h"
@@ -14,6 +15,8 @@
 namespace mozilla {
 namespace dom {
 
+class WindowGlobalParent;
+
 #define MOZ_EACH_WC_FIELD(FIELD)                                       \
   FIELD(OuterWindowId, uint64_t)                                       \
   FIELD(CookieJarSettings, Maybe<mozilla::net::CookieJarSettingsArgs>) \
@@ -36,6 +39,10 @@
   uint64_t OuterWindowId() const { return GetOuterWindowId(); }
   bool IsDiscarded() const { return mIsDiscarded; }
 
+  bool IsCached() const;
+
+  Span<RefPtr<BrowsingContext>> Children() { return mChildren; }
+
   // Cast this object to it's parent-process canonical form.
   WindowGlobalParent* Canonical();
 
@@ -50,6 +57,12 @@
     uint64_t mBrowsingContextId;
 
     FieldTuple mFields;
+
+    bool operator==(const IPCInitializer& aOther) const {
+      return mInnerWindowId == aOther.mInnerWindowId &&
+             mBrowsingContextId == aOther.mBrowsingContextId &&
+             mFields == aOther.mFields;
+    }
   };
   IPCInitializer GetIPCInitializer() {
     return {mInnerWindowId, mBrowsingContext->Id(), mFields.Fields()};
@@ -66,6 +79,11 @@
   void Init();
 
  private:
+  friend class BrowsingContext;
+
+  void AppendChildBrowsingContext(BrowsingContext* aBrowsingContext);
+  void RemoveChildBrowsingContext(BrowsingContext* aBrowsingContext);
+
   // Send a given `BaseTransaction` object to the correct remote.
   void SendCommitTransaction(ContentParent* aParent,
                              const BaseTransaction& aTxn, uint64_t aEpoch);
@@ -100,6 +118,12 @@
   uint64_t mInnerWindowId;
   RefPtr<BrowsingContext> mBrowsingContext;
 
+  // --- NEVER CHANGE `mChildren` DIRECTLY! ---
+  // Changes to this list need to be synchronized to the list within our
+  // `mBrowsingContext`, and should only be performed through the
+  // `AppendChildBrowsingContext` and `RemoveChildBrowsingContext` methods.
+  nsTArray<RefPtr<BrowsingContext>> mChildren;
+
   bool mIsDiscarded = false;
 };
 